fix: refactored tests dependant on float assertion

jaredfolkins opened the pull request:
Hi Dan,

As promised, here are the fixes for the tests. 

I created some helper methods which essentially ripoff Go's stdlib.

From there I refactored the coordinates into float64 vertices. 

I then used a helper method I created to compare the coordinates instantiated to the Lines and Points returned, throwing an error if the deviation is too great.

I think this is readable but am open to feedback.

Jared

jaredfolkins committed:
fix: refactored tests dependant on float assertion

dancannon commented:
Thanks, ill give it a read through. 

Ignore the failing tests that is another issue that is unrelated. Ill fix that in master before merging.

jaredfolkins commented:
@dancannon Dude. I needed the payoff of passing GREEN tests. I needed it so bad. :-) #testJunkie

dancannon commented:
Ah I had fixed the issue but I must have incorrectly resolved a merge conflict. Ill fix it and the tests should be green. Finally... :D

dancannon commented:
Soon!

jaredfolkins commented:
:+1:

dancannon commented on query_geospatial_test.go:
```diff
@@ -1,42 +1,89 @@
+	for _, points := range lines {
+		for ip, point := range points {
+			v := reflect.ValueOf(point)
+			for i := 0; i < v.NumField(); i++ {
```
While it doesnt make too much difference I think it would probably be clearer if we avoided using reflection here.

jaredfolkins replied:
Oh goody. Maybe you can teach me something. How do I iterate over a Struct's fields without using reflection? (No snark btw. I really am curious if I'm missing something)

dancannon replied:
Sadly you cant :) I was just think of doing something like this (please tell me if I have missed something). Again it should make any real difference but I try to avoid magic where possible. What do you think?

``` go
if !kindaclose(co[ip][0], point.Lon) {
    c.Errorf("the deviation between the compared floats is too great [%v:%v]", co[ip][0], point.Lon)
}
if !kindaclose(co[ip][1], point.Lat) {
    c.Errorf("the deviation between the compared floats is too great [%v:%v]", co[ip][1], point.Lat)
}
```

jaredfolkins replied:
Hah. Much simpler. But in that case I'd suggest the following for readability and error reporting.

``` go
func compareCoordinates(co [][]float64, lines types.Lines, c *test.C) {
    for _, points := range lines {
        for i, point := range points {
            lon := co[i][0]
            lat := co[i][1]
            if !kindaclose(lon, point.Lon) {
                c.Errorf("the deviation between the compared floats is too great [%v:%v]", lon, point.Lon)
            }
            if !kindaclose(lat, point.Lat) {
                c.Errorf("the deviation between the compared floats is too great [%v:%v]", lat, point.Lat)
            }
        }
    }
}
```

dancannon replied:
Even better.

dancannon commented:
New update, I think the tests were fine but there might be an issue with the build process. For whatever reason the workspace was not properly updated to the correct commit. Tests are green now anyway.

With that over time to review...
1. Is there any reason why you chose to only use the new `compareCoordinates` function on some of the tests.
2. Do we need any of the comparison functions other than `kindaclose`?

jaredfolkins commented:
1. I refactored only the tests that were failing for me. I agree that using the helper functions for all applicable tests is ideal but wanted to make sure I was creating solid code for you before refactoring further.
2. I added all the helper functions because I found them helpful. It was hard to tell going forward what precision would be needed when testing floats. It won't offend me one bit if you want me to remove any of them.

dancannon commented:
1. Makes sense, I think it would make sense to migrate all geospatial tests to your new test format. I can make that change if you want?
2. I dont really have anything against the helper functions, we might as well keep them for now as they are not exactly difficult to maintain

jaredfolkins commented:
1. @dancannon refactor away! :+1: 
2. You had mentioned that you were going to utilize a new testing package. I'd keep the helpers in there until you make the move. Also, how do you test just a certain test? The following gets eaten and running all the tests costs me time.

``` bash
go test -run=SomeTestName -v
```

dancannon commented:
1. I was just thinking that, the helpers are staying in then!

Because the driver uses gocheck (unfortunately) you need to run `go test -gocheck=SomeTestName`. When I eventually switch to http://github.com/stretchr/testify these issues should go away.

jaredfolkins commented:
A small aside but what does a testing framework get you? I only ask because including one is another barrier to contributions. I'm generally interested in the pros/cons in your opinion.

dancannon commented:
Honestly I am planning on using testify just for its assertions.  I dont want to replace the `go test` behaviour again.

jaredfolkins committed:
enhc: added benchmarks, stubbed TestMain(), Setup(), Teardown()

jaredfolkins commented:
Hi Dan,

As discussed here are the write benchmarks. 

I added some stubs also to gorethink_test.go for Setup and Teardown. 

I also added to the README, obviously feel free to refactor the wording for what works for you. I just put my best guess into how it should sound.

Later,
Jared

dancannon commented:
Merged to master, thanks for your help @jaredfolkins. I hope you dont mind but I changed the README slightly.

jaredfolkins commented:
@dancannon Hah. Thanks for the shoutout. Not needed but very much appreciated.
